###TITANIC SOLUÇÃO
 
 
Em 15 de abril de 1912, durante sua viagem inaugural, o Titanic afundou após colidir com um iceberg, matando 1502 de 2224 passageiros e tripulantes. Taxa de sobrevivência de 32%. Uma das razões pelas quais o naufrágio levou a tal perda de vidas foi que não havia botes salva vidas suficientes para os passageiros e a tripulação. Embora houvesse algum elemento de sorte envolvido na sobrevivência do naufrágio, alguns grupos de pessoas eram mais propensos a sobreviver do que outros, como mulheres, crianças e a classe alta

####OBJETIVO
 
Esse notebook tem como objetivo mostrar o passo a passo de seu desenvolvimento explicando e justificando cada decisão tomada durante o desenvolvimento da solução.
 
Dessa forma definiremos sete etapas para alcançar nosso objetivo.
 
1.Definição do problema.
 
2.Obtenção dos dados
 
3.Exploração dos dados
 
4.Preparar, limpar os dados.
 
5.Analisar, identificar  padrões e explorar os dados.
 
6.Modelar, prever e resolver o problema.
 
7.Avaliação e resultados.

###PROBLEMA
Quais passageiros  sobreviveram ou não ao desastre do Titanic?

###ESTRATÉGIAS
 
 
Podemos classificar ou categorizar nossas amostras. Também podemos entender as implicações ou correlação de diferentes classes com nosso objetivo de solução.
 
Abordarademos a correlação das features disponíveis no conjunto de dados de treinamento. Quais features do conjunto de dados contribuem significativamente para nossa solução
 
Para o estágio de modelagem, é necessário preparar o conjunto de dados. Dependendo da escolha do algoritmo do modelo, pode-se exigir que todos os sejam convertidos em valores numéricos equivalentes. Por exemplo, convertendo valores categóricos de texto em valores numéricos.
 
A preparação de dados também pode exigir que calculemos quaisquer valores faltantes das features. Os algoritmos de modelo podem funcionar melhor quando não há valores faltantes.
 
Iremos analisar o conjunto de dados de treinamento fornecido para erros ou valores possivelmente imprecisos nas features e tentar corrigir esses valores ou excluir as amostras que contêm os erros. Uma maneira de fazer isso é detectar discrepâncias entre nossas amostras ou recursos. Também podemos descartar completamente uma features se ela não estiver contribuindo para a análise ou se ela esta distorcendo  significativamente os resultados.
 
Podemos criar novas features com base em outra features existente ou em um conjunto de features.

###Dicionário de Dados
 
1.PassengerId: Número de identificação do passageiro
 
2.Survived: Informa se o passageiro sobreviveu ao desastre 0 = Não, 1 = Sim
 
3.Pclass: Classe do bilhete 1 = 1ª Classe, 2 = 2ª Classe, 3 = 3ª Classe
 
4.Name: Nome do passageiro
 
5.Sex: Sexo do passageiro
 
6.Age: Idade do passageiro
 
7.SibSp: Quantidade de cônjuges e irmãos a bordo 
8.Parch: Quantidade de pais e filhos a bordo
 
9.Ticket: Número da passagem
 
10.Fare: Preço da Passagem
 
11.Cabin: Número da cabine do passageiro
 
12.Embarked: Porto no qual o passageiro embarcou C = Cherbourg, Q = Queenstown, S = Southampton

Montando o drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Importando as bibliotecas básicas

In [ ]:
# data analysis and wrangling
import pandas as pd
import numpy as np
import random as rnd
 
# visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
 
# machine learning
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC, LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


Carregando a Base de Dados

In [ ]:
train_df=pd.read_csv('/content/drive/My Drive/Colab Notebooks/Titanic kaggle/train.csv')
test_df=pd.read_csv('/content/drive/My Drive/Colab Notebooks/Titanic kaggle/test.csv')
combine = [train_df, test_df]

  ## Analisando e descrevendo dados 
A biblioteca Pandas irá nos ajudar a descrever os conjuntos de dados que sera fundamental para mostrar as features disponíveis no conjunto de dados.

In [ ]:
print(train_df.columns.values)

['PassengerId' 'Survived' 'Pclass' 'Name' 'Sex' 'Age' 'SibSp' 'Parch'
 'Ticket' 'Fare' 'Cabin' 'Embarked']


** Quais features são categóricas? ** 
 
 
- Categórico: Survived, Sex, and Embarked. Ordinal: Pclass.
 
** Quais features são numéricas? **
 
 
- numéricas: Age, Fare. Discrete: SibSp, Parch.

In [ ]:
# preview the data
train_df.head()

  ** Quais features são tipos de dados mistos? ** 
- Ticket é uma mistura de tipos de dados numéricos e alfanuméricos. A cabine é alfanumérica.
 
** Quais features podem conter erros ou erros de digitação? **
 
- A feature name pode conter erros ou erros de digitação, pois existem várias maneiras de descrever um nome, incluindo títulos, colchetes e aspas usadas para nomes alternativos ou abreviados.

In [ ]:
train_df.tail()

PassengerId  Survived  Pclass  ...   Fare Cabin  Embarked
886          887         0       2  ...  13.00   NaN         S
887          888         1       1  ...  30.00   B42         S
888          889         0       3  ...  23.45   NaN         S
889          890         1       1  ...  30.00  C148         C
890          891         0       3  ...   7.75   NaN         Q

[5 rows x 12 columns]

  ** Quais features contêm valores em branco, nulos ou vazios? ** 
Isso exigirá correção.
 
- Cabin> Age> Embarked são features que contêm vários valores nulos nessa ordem para o conjunto de dados de treinamento.
- Cabin> Age incompleta no caso de um conjunto de dados de teste.
 
** Quais são os tipos de dados presente nas features? **
 
 
- Sete features são números inteiros ou flutuantes. Seis no caso do conjunto de dados de teste.
- Cinco features são strings (objeto).

In [ ]:
train_df.info()
print('_'*40)
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
________________________________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Passenger

  ** Qual é a distribuição dos valores das features numéricos entre as amostras? ** 
 
- O total de amostras corresponde a 891 ou 40% do número real de passageiros a bordo do Titanic (2.224).
- Survived é uma feature categórica com valores 0 ou 1.
- Cerca de 38% das amostras sobreviveram, representativas da taxa de sobrevivência real em 32%.
- A maioria dos passageiros (> 75%) não viajou com pais ou filhos.
- Quase 30% dos passageiros tinham irmãos e / ou cônjuge a bordo.
- As tarifas variaram significativamente, com poucos passageiros (<1%) pagando até US $ 512.
- Poucos passageiros idosos (<1%) na faixa etária de 65 a 80 anos.

In [ ]:
train_df.describe()

  ** Qual é a distribuição das features  categóricos? ** 
- Os nomes são únicos no conjunto de dados (count = unique = 891)
- Variável sexo como dois valores possíveis com 65% do sexo masculino (top = masculino, freq = 577 / contagem = 891).
- Os valores da cabin têm vários duplicados  nas amostras. Provavelmente , vários passageiros compartilhavam uma cabine.
- Embarked tem três valores possíveis. "S" usada pela maioria dos passageiros
- A feature ticket possui uma alta taxa (22%) de valores duplicados (exclusivo = 681).

In [ ]:
train_df.describe(include=['O'])

Name   Sex  Ticket Cabin Embarked
count                         891   891     891   204      889
unique                        891     2     681   147        3
top     Danbom, Mr. Ernst Gilbert  male  347082    G6        S
freq                            1   577       7     4      644

  ### Prognóstico baseados na análise de dados 
Apartir da análise  de dados  feita chegamos às seguintes premissas.
 
** Correlação. **
 
Queremos saber até que ponto cada feature se correlaciona com a Survival. É importante verificar o nível de correlação para que no inicio do projeto possamos fazer combinações. 
 
** Conclusões. **
 
1.A feature Age está definitivamente relacionado à sobrevivência.
2.A feature Embarked, pode estar relacionado à sobrevivência ou a outra feature importante.
 
** Corrigir. **
 
1. A feature ticket pode ser retirada de nossa análise, pois contém uma alta taxa de duplicação (22%) e pode não haver uma correlação entre o ticket e a suvive.
2. A feature cabine pode ser descartada, pois é altamente incompleta ou contém muitos valores nulos no conjunto de dados de treinamento e teste.
3. O PassengerId pode ser retirado do conjunto de dados de treinamento, pois não contribui para a sobrevivência.
4. A feature nome é relativamente fora do padrão, pode não contribuir diretamente para a sobrevivência, então talvez seja descartado.
 
** Criando. **
 
1. Podemos criar uma nova feature chamada Família, baseado em Parch e SibSp, para obter a contagem total de membros da família a bordo.
2. Podemos utilizar a feature Name para extrair o Título como uma nova feature.
3. Podemos criar uma nova feature para faixas etárias. Isso transforma um recurso numérico contínuo em um recurso categórico ordinal.
4. Também podemos criar uma feature de faixa de tarifa se isso ajudar nossa análise.
 
** Classificação. **
 
Também podemos adicionar nossas suposições com base na descrição do problema observada anteriormente.
 
1. As mulheres (sexo = mulher) tiveram maior probabilidade de sobreviver.
2. As crianças (idade <?) Tiveram maior probabilidade de sobreviver.
3. Os passageiros da classe alta (Pclass = 1) tiveram maior probabilidade de sobreviver.

In [ ]:
train_df[['Pclass', 'Survived']].groupby(['Pclass'], as_index=False).mean().sort_values(by='Survived', ascending=False)

In [ ]:
 train_df[["Sex", "Survived"]].groupby(['Sex'], as_index=False).mean().sort_values(by='Survived', ascending=False)

In [ ]:
train_df[["Sex",'Pclass', "Survived"]].groupby(['Sex','Pclass'], as_index=False).mean().sort_values(by='Survived', ascending=False)

In [ ]:
train_df[["SibSp", "Survived"]].groupby(['SibSp'], as_index=False).mean().sort_values(by='Survived', ascending=False)

In [ ]:
train_df[["Parch", "Survived"]].groupby(['Parch'], as_index=False).mean().sort_values(by='Survived', ascending=False)

 ## Analise visualizando dados 
Agora podemos continuar confirmando algumas de nossas suposições usando visualizações para analisar os dados.
 
### Correlacionando features numéricas
 
Vamos começar entendendo as correlações entre os recursos numéricos e nossa meta de solução (Survived).
 
Um gráfico de histograma é útil para analisar variáveis ​​numéricas contínuas, como Idade, em que faixa ajudarão a identificar padrões úteis.
 
** Observações. **
 
- Bebês (idade <= 4) tiveram alta taxa de sobrevivência.
- Passageiros mais velhos (idade = 80) sobreviveram.
- Grande número de jovens de 15 a 25 anos não sobreviveu.
- A maioria dos passageiros tem entre 15 e 35 anos.
 
** Decisões. **
 
Apartir dessa análise devemos considerar a idade em nosso modelo de treinamento. Assim devemos:
 
- Completar a feature Age para valores nulos.
- Agrupar faixas etárias a feature Age.

In [ ]:
g = sns.FacetGrid(train_df, col='Survived')
g.map(plt.hist, 'Age', bins=40)

 ### Correlacionando features numéricas e ordinais 
Podemos combinar vários recursos para identificar correlações usando um único gráfico. Isso pode ser feito com recursos numéricos e categóricos que possuem valores numéricos.
 
** Observações. **
 
- Pclass = 3 teve a maioria dos passageiros, porém a maioria não sobreviveu. Confirma nossa suposição de classificação nº 2.
- Os passageiros infantis da Pclass = 2 e Pclass = 3 sobreviveram principalmente. Qualifica ainda mais nossa suposição classificativa nº 2.
- A maioria dos passageiros da Pclass = 1 sobreviveu. Confirma nossa suposição de classificação nº 3.
- Pclass varia em termos de distribuição etária dos passageiros.
 
** Decisões. **
 
- Considere Pclass para o treinamento do modelo.

In [ ]:
# grid = sns.FacetGrid(train_df, col='Pclass', hue='Survived')
grid = sns.FacetGrid(train_df, col='Survived', row='Pclass', size=2.2, aspect=1.6)
grid.map(plt.hist, 'Age', alpha=.5, bins=20)
grid.add_legend();

 ### Correlacionando features categóricos 
Agora podemos correlacionar recursos categóricos com nosso objetivo de solução.
 
** Observações. **
 
- As passageiras tiveram uma taxa de sobrevivência muito melhor que os homens. Confirma a classificação (# 1).
- Exceção em Embarked = C, onde os homens tiveram maior taxa de sobrevivência. Esta poderia ser uma correlação entre Pclass e Embarked e, por sua vez, Pclass e Survived, não necessariamente correlação direta entre Embarked e Survived.
- Os machos apresentaram melhor taxa de sobrevivência na Pclass = 3 quando comparados à Pclass = 2 nos portos C e Q. Concluindo (nº 2).
- Os portos de embarque têm taxas de sobrevivência variadas para Pclass = 3 e entre passageiros do sexo masculino. Correlação (# 1).
 
** Decisões. **
 
- Adicione o recurso Sexo ao modelo de treinamento.
- Complete e adicione o recurso Embarked ao modelo de treinamento.

In [ ]:
# grid = sns.FacetGrid(train_df, col='Embarked')
grid = sns.FacetGrid(train_df, row='Embarked', size=2.2, aspect=1.6)
grid.map(sns.pointplot, 'Pclass', 'Survived', 'Sex', palette='deep')
grid.add_legend()

 ### Correlacionando features categóricos e numéricos 
Também podemos correlacionar features categóricos (com valores não numéricos) e recursos numéricos. Podemos considerar correlacionar Embarked (categórico não numérico), Sexo (categórico não numérico), Tarifa (contínuo numérico), com Survived (categórico numérico).
 
** Observações. **
 
- Passageiros com tarifa mais alta tiveram melhor sobrevida. Confirma nossa suposição para a criação de (# 4) intervalos de tarifas.
- O porto de embarque se correlaciona com as taxas de sobrevivência. Confirma a correlação (nº 1) e a conclusão (nº 2).

In [ ]:
# grid = sns.FacetGrid(train_df, col='Embarked', hue='Survived', palette={0: 'k', 1: 'w'})
grid = sns.FacetGrid(train_df, row='Embarked', col='Survived', size=2.2, aspect=1.6)
grid.map(sns.barplot, 'Sex', 'Fare', alpha=.5, ci=None)
grid.add_legend()

 ## Organizar dados 
Reunimos várias suposições e decisões sobre nossos conjuntos de dados. Agora vamos agora executar nossas decisões e suposições para corrigir, criar e concluir nossas metas.
 
### Corrigindo e descartando recursos
Com base em nossas suposições e decisões, iremos descartar os recursos de cabine e ticket.

In [ ]:
print("Before", train_df.shape, test_df.shape, combine[0].shape, combine[1].shape)
 
train_df = train_df.drop(['Ticket', 'Cabin'], axis=1)
test_df = test_df.drop(['Ticket', 'Cabin'], axis=1)
combine = [train_df, test_df]
 
"After", train_df.shape, test_df.shape, combine[0].shape, combine[1].shape

Before (891, 12) (418, 11) (891, 12) (418, 11)


('After', (891, 10), (418, 9), (891, 10), (418, 9))

  ### Criando uma nova features a partir  dos existentes 
Queremos analisar se a features Name pode ser utilizada para extrair títulos e testar a correlação entre títulos e sobrevivência, antes de descartar a features Name e PassengerId.
 
No código a seguir, extraímos a features Título usando expressões regulares. O padrão RegEx `(\ w + \.)` Corresponde à primeira palavra que termina com um caractere de ponto no recurso Name. O sinalizador `expand = False` retorna um DataFrame.
 
** Observações. **
 
Quando plotamos Título, Idade e Sobrevividos, observamos as seguintes observações.
 
- A maioria dos títulos agrupa faixas etárias com precisão. Por exemplo: o título de mestre tem uma média de idade de 5 anos.
- A sobrevivência entre as faixas etárias do título varia um pouco.
- Certos títulos sobreviveram principalmente (Mme, Lady, Sir) ou não (Don, Rev, Jonkheer).
 
**Decisão.**
 
- Decidimos manter o novo recurso Título para o treinamento do modelo.

In [ ]:
for dataset in combine:
    dataset['Title'] = dataset.Name.str.extract(' ([A-Za-z]+)\.', expand=False)
 
pd.crosstab(train_df['Title'], train_df['Sex'])

Sex       female  male
Title                 
Capt           0     1
Col            0     2
Countess       1     0
Don            0     1
Dr             1     6
Jonkheer       0     1
Lady           1     0
Major          0     2
Master         0    40
Miss         182     0
Mlle           2     0
Mme            1     0
Mr             0   517
Mrs          125     0
Ms             1     0
Rev            0     6
Sir            0     1

  
Podemos substituir muitos títulos por um nome mais comum ou classificá-los como `Raros`.

In [ ]:
for dataset in combine:
    dataset['Title'] = dataset['Title'].replace(['Lady', 'Countess','Capt', 'Col',\
    'Don', 'Dr', 'Major', 'Rev', 'Sir', 'Jonkheer', 'Dona'], 'Rare')
 
    dataset['Title'] = dataset['Title'].replace('Mlle', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Ms', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Mme', 'Mrs')
    
train_df[['Title', 'Survived']].groupby(['Title'], as_index=False).mean()

Title  Survived
0  Master  0.575000
1    Miss  0.702703
2      Mr  0.156673
3     Mrs  0.793651
4    Rare  0.347826

 Podemos converter os títulos categóricos em ordinais.

In [ ]:
title_mapping = {"Mr": 1, "Miss": 2, "Mrs": 3, "Master": 4, "Rare": 5}
for dataset in combine:
    dataset['Title'] = dataset['Title'].map(title_mapping)
    dataset['Title'] = dataset['Title'].fillna(0)
 
train_df.head()

PassengerId  Survived  Pclass  ...     Fare Embarked  Title
0            1         0       3  ...   7.2500        S      1
1            2         1       1  ...  71.2833        C      3
2            3         1       3  ...   7.9250        S      2
3            4         1       1  ...  53.1000        S      3
4            5         0       3  ...   8.0500        S      1

[5 rows x 11 columns]

  Agora, podemos descartar com segurança o recurso Name dos conjuntos de dados de treinamento e teste. Também não precisamos do recurso PassengerId no conjunto de dados de treinamento.

In [ ]:
train_df = train_df.drop(['Name', 'PassengerId'], axis=1)
test_df = test_df.drop(['Name'], axis=1)
combine = [train_df, test_df]
train_df.shape, test_df.shape

((891, 9), (418, 9))

  ### Convertendo uma feature categórico
 
Vamos começar convertendo o recurso Sexo para um novo recurso chamado Gênero, onde feminino = 1 e masculino = 0.

In [ ]:
for dataset in combine:
    dataset['Sex'] = dataset['Sex'].map( {'female': 1, 'male': 0} ).astype(int)
 
train_df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked  Title
0         0       3    0  22.0      1      0   7.2500        S      1
1         1       1    1  38.0      1      0  71.2833        C      3
2         1       3    1  26.0      0      0   7.9250        S      2
3         1       1    1  35.0      1      0  53.1000        S      3
4         0       3    0  35.0      0      0   8.0500        S      1

  ### Concluindo um recurso numérico contínuo 
Agora devemos começar a estimar e concluir os recursos com valores ausentes ou nulos. Primeiro, faremos isso para o recurso Age.

In [ ]:
# grid = sns.FacetGrid(train_df, col='Pclass', hue='Gender')
grid = sns.FacetGrid(train_df, row='Pclass', col='Sex', size=2.2, aspect=1.6)
grid.map(plt.hist, 'Age', alpha=.5, bins=20)
grid.add_legend()

  Vamos começar preparando uma matriz vazia para conter valores da idade calculados com base nas combinações Pclass x Gênero.

In [ ]:
guess_ages = np.zeros((2,3))
guess_ages

  Agora, iteramos sobre Sexo (0 ou 1) e Pclass (1, 2, 3) para calcular os valores calculados da idade para as seis combinações.

In [ ]:
for dataset in combine:
    for i in range(0, 2):
        for j in range(0, 3):
            guess_df = dataset[(dataset['Sex'] == i) & \
                                  (dataset['Pclass'] == j+1)]['Age'].dropna()
 
            # age_mean = guess_df.mean()
            # age_std = guess_df.std()
            # age_guess = rnd.uniform(age_mean - age_std, age_mean + age_std)
 
            age_guess = guess_df.median()
 
            # Convert random age float to nearest .5 age
            guess_ages[i,j] = int( age_guess/0.5 + 0.5 ) * 0.5
            
    for i in range(0, 2):
        for j in range(0, 3):
            dataset.loc[ (dataset.Age.isnull()) & (dataset.Sex == i) & (dataset.Pclass == j+1),\
                    'Age'] = guess_ages[i,j]
 
    dataset['Age'] = dataset['Age'].astype(int)
 
train_df.head()

NameError: ignored

  Vamos criar faixas etárias e determinar correlações com Survived.

In [ ]:
train_df['AgeBand'] = pd.cut(train_df['Age'], 5)
train_df[['AgeBand', 'Survived']].groupby(['AgeBand'], as_index=False).mean().sort_values(by='AgeBand', ascending=True)

In [ ]:
train_df['AgeBand'] = pd.cut(train_df['Age'], 5)
train_df[['AgeBand','Sex', 'Survived']].groupby(['AgeBand','Sex'], as_index=False).mean().sort_values(by='AgeBand', ascending=True)

  Vamos substituir Age por ordinais com base nessas bandas.

In [ ]:
for dataset in combine:    
    dataset.loc[ dataset['Age'] <= 16, 'Age'] = 0
    dataset.loc[(dataset['Age'] > 16) & (dataset['Age'] <= 32), 'Age'] = 1
    dataset.loc[(dataset['Age'] > 32) & (dataset['Age'] <= 48), 'Age'] = 2
    dataset.loc[(dataset['Age'] > 48) & (dataset['Age'] <= 64), 'Age'] = 3
    dataset.loc[ dataset['Age'] > 64, 'Age']
train_df.head()

Survived  Pclass  Sex  Age  SibSp  Parch     Fare Embarked  Title
0         0       3    0  1.0      1      0   7.2500        S      1
1         1       1    1  2.0      1      0  71.2833        C      3
2         1       3    1  1.0      0      0   7.9250        S      2
3         1       1    1  2.0      1      0  53.1000        S      3
4         0       3    0  2.0      0      0   8.0500        S      1

Podemos remover a feature AgeBand.

In [ ]:
train_df = train_df.drop(['AgeBand'], axis=1)
combine = [train_df, test_df]
train_df.head()

In [ ]:
for dataset in combine:
    dataset['FamilySize'] = dataset['SibSp'] + dataset['Parch'] + 1
 
train_df[['FamilySize', 'Survived']].groupby(['FamilySize'], as_index=False).mean().sort_values(by='Survived', ascending=False)

In [ ]:
for dataset in combine:
    dataset['FamilySize'] = dataset['SibSp'] + dataset['Parch'] + 1
 
train_df[['FamilySize','Sex', 'Survived']].groupby(['FamilySize','Sex'], as_index=False).mean().sort_values(by='Survived', ascending=False)

  ### Criando uma nova features combinando recursos existentes 
Podemos criar uma nova features para o FamilySize que combina Parch e SibSp. Isso nos permitirá remover Parch e SibSp de nossos conjuntos de dados.

  Podemos criar outra feature chamado IsAlone.

In [ ]:
for dataset in combine:
    dataset['IsAlone'] = 0
    dataset.loc[dataset['FamilySize'] == 1, 'IsAlone'] = 1
 
train_df[['IsAlone', 'Survived']].groupby(['IsAlone'], as_index=False).mean()

  Vamos deixar de lado as features Parch, SibSp e FamilySize em favor do IsAlone.

In [ ]:
train_df = train_df.drop(['Parch', 'SibSp', 'FamilySize'], axis=1)
test_df = test_df.drop(['Parch', 'SibSp', 'FamilySize'], axis=1)
combine = [train_df, test_df]
 
train_df.head()

  Também podemos criar uma feature artificial combinando Pclass e Age.

In [ ]:
for dataset in combine:
    dataset['Age*Class'] = dataset.Age * dataset.Pclass
 
train_df.loc[:, ['Age*Class', 'Age', 'Pclass']].head(10)

  ### Concluindo um recurso categórico 
O recurso embarcado aceita valores S, Q, C com base no porto de embarque. Nosso conjunto de dados de treinamento tem dois valores ausentes. Nós simplesmente os preenchemos com a ocorrência mais comum.

In [ ]:
freq_port = train_df.Embarked.dropna().mode()[0]
freq_port

In [ ]:
for dataset in combine:
    dataset['Embarked'] = dataset['Embarked'].fillna(freq_port)
    
train_df[['Embarked', 'Survived']].groupby(['Embarked'], as_index=False).mean().sort_values(by='Survived', ascending=False)

In [ ]:
 
for dataset in combine:
    dataset['Embarked'] = dataset['Embarked'].fillna(freq_port)
    
train_df[['Embarked','Sex', 'Survived']].groupby(['Embarked','Sex'], as_index=False).mean().sort_values(by='Survived', ascending=False)

  ### Convertendo feature categórico em numérico 
Agora podemos converter a feature EmbarkedFill criando uma nova feature numérica.

In [ ]:
for dataset in combine:
    dataset['Embarked'] = dataset['Embarked'].map( {'S': 0, 'C': 1, 'Q': 2} ).astype(int)
 
train_df.head()

  ### Concluindo e convertendo rapidamente um recurso numérico 
Agora podemos concluir o recurso Tarifa no qual possue um único valor ausente no conjunto de dados de teste desse modo iremos substituir pelo o valor que ocorre com mais freqüência nesse recurso.
Também podemos arredondar a tarifa para duas casas decimais, pois representa a moeda.

In [ ]:
test_df['Fare'].fillna(test_df['Fare'].dropna().median(), inplace=True)
test_df.head()

Podemos criar o FareBand.

In [ ]:
train_df['FareBand'] = pd.qcut(train_df['Fare'], 4)
train_df[['FareBand', 'Survived']].groupby(['FareBand'], as_index=False).mean().sort_values(by='FareBand', ascending=True)

In [ ]:
 
train_df['FareBand'] = pd.qcut(train_df['Fare'], 4)
train_df[['FareBand','Sex', 'Survived']].groupby(['FareBand','Sex'], as_index=False).mean().sort_values(by='FareBand', ascending=True)

  Convertendo a feature Fare em valores ordinais com base no FareBand.

In [ ]:
for dataset in combine:
    dataset.loc[ dataset['Fare'] <= 7.91, 'Fare'] = 0
    dataset.loc[(dataset['Fare'] > 7.91) & (dataset['Fare'] <= 14.454), 'Fare'] = 1
    dataset.loc[(dataset['Fare'] > 14.454) & (dataset['Fare'] <= 31), 'Fare']   = 2
    dataset.loc[ dataset['Fare'] > 31, 'Fare'] = 3
    dataset['Fare'] = dataset['Fare'].astype(int)
 
train_df = train_df.drop(['FareBand'], axis=1)
combine = [train_df, test_df]
    
train_df.head(10)

  E o conjunto de dados de teste.

In [ ]:
test_df.head(10)

  Apos modelar, prever e resolver, estamos prontos para treinar um modelo e prever a solução necessária. Nosso problema é um problema de classificação e regressão. Queremos identificar a relação entre a saída (Sobrevivido ou não) com outras variáveis ​​ou características (Gênero, Idade, Porto ...). Também estamos desenvolvendo uma categoria de aprendizado de máquina chamada aprendizado supervisionado, enquanto treinamos nosso modelo com um determinado conjunto de dados. Com esses dois critérios - Aprendizado supervisionado e Classificação e regressão, podemos restringir nossa escolha do nosso modelo.
 
- Regressão logística
- KNN ou k-vizinhos mais próximos
- Máquinas de vetores de suporte
- Classificador Naive Bayes
- Árvore de decisão
- Random Forrest
- Perceptron
- Rede neural artificial
- RVM ou Máquina de Vetor de Relevância

In [ ]:
X_train = train_df.drop("Survived", axis=1)
Y_train = train_df["Survived"]
X_test  = test_df.drop("PassengerId", axis=1).copy()
X_train.shape, Y_train.shape, X_test.shape

  A regressão logística é um modelo útil para ser executado no início do fluxo de trabalho. A regressão logística mede o relacionamento entre a variável dependente categórica (característica) e uma ou mais variáveis ​​independentes (características) estimando probabilidades usando uma função logística, que é a distribuição logística cumulativa.

In [ ]:
# Logistic Regression
 
logreg = LogisticRegression()
logreg.fit(X_train, Y_train)
Y_pred = logreg.predict(X_test)
acc_log = round(logreg.score(X_train, Y_train) * 100, 2)
acc_log

In [ ]:
coeff_df = pd.DataFrame(train_df.columns.delete(0))
coeff_df.columns = ['Feature']
coeff_df["Correlation"] = pd.Series(logreg.coef_[0])
 
coeff_df.sort_values(by='Correlation', ascending=False)

  A seguir, modelamos usando Máquinas de Vetor de Suporte, que são modelos de aprendizado supervisionado com algoritmos de aprendizado associados que analisam dados usados ​​para análise de classificação e regressão. Dado um conjunto de amostras de treinamento, cada uma marcada como pertencente a uma ou outra de ** duas categorias **, um algoritmo de treinamento SVM constrói um modelo que atribui novas amostras de teste a uma categoria ou a outra, tornando-o um binário não probabilístico classificador linear.

In [ ]:
# Support Vector Machines
 
svc = SVC()
svc.fit(X_train, Y_train)
Y_pred = svc.predict(X_test)
acc_svc = round(svc.score(X_train, Y_train) * 100, 2)
acc_svc

  No reconhecimento de padrões, o algoritmo k-vizinhos mais próximos (ou k-NN) é um método não paramétrico usado para classificação e regressão. Uma amostra é classificada pelo voto da maioria de seus vizinhos, sendo a amostra atribuída à classe mais comum entre os k vizinhos mais próximos (k é um número inteiro positivo, geralmente pequeno). Se k = 1, o objeto é simplesmente atribuído à classe do único vizinho mais próximo.

In [ ]:
knn = KNeighborsClassifier(n_neighbors = 3)
knn.fit(X_train, Y_train)
Y_pred = knn.predict(X_test)
acc_knn = round(knn.score(X_train, Y_train) * 100, 2)
acc_knn

  No aprendizado de máquina, os classificadores ingênuos de Bayes são uma família de classificadores probabilísticos simples, baseados na aplicação do teorema de Bayes com fortes (ingênuas) suposições de independência entre os recursos. Os classificadores Naive Bayes são altamente escaláveis, exigindo vários parâmetros lineares no número de variáveis ​​(características) em um problema de aprendizagem.

In [ ]:
# Gaussian Naive Bayes
 
gaussian = GaussianNB()
gaussian.fit(X_train, Y_train)
Y_pred = gaussian.predict(X_test)
acc_gaussian = round(gaussian.score(X_train, Y_train) * 100, 2)
acc_gaussian

 O perceptron é um algoritmo para aprendizado supervisionado de classificadores binários (funções que podem decidir se uma entrada, representada por um vetor de números, pertence a alguma classe específica ou não). É um tipo de classificador linear, isto é, um algoritmo de classificação que faz suas previsões com base em uma função preditora linear que combina um conjunto de pesos com o vetor de característica. O algoritmo permite o aprendizado on-line, na medida em que processa elementos no conjunto de treinamento, um de cada vez.

In [ ]:
# Perceptron
 
perceptron = Perceptron()
perceptron.fit(X_train, Y_train)
Y_pred = perceptron.predict(X_test)
acc_perceptron = round(perceptron.score(X_train, Y_train) * 100, 2)
acc_perceptron

In [ ]:
# Linear SVC
 
linear_svc = LinearSVC()
linear_svc.fit(X_train, Y_train)
Y_pred = linear_svc.predict(X_test)
acc_linear_svc = round(linear_svc.score(X_train, Y_train) * 100, 2)
acc_linear_svc

In [ ]:
# Stochastic Gradient Descent
 
sgd = SGDClassifier()
sgd.fit(X_train, Y_train)
Y_pred = sgd.predict(X_test)
acc_sgd = round(sgd.score(X_train, Y_train) * 100, 2)
acc_sgd

 Esse modelo usa uma árvore de decisão como modelo preditivo, que mapeia recursos (galhos de árvores) para conclusões sobre o valor alvo (folhas de árvores). Os modelos de árvore em que a variável de destino pode receber um conjunto finito de valores são chamados de árvores de classificação; nessas estruturas em árvore, as folhas representam rótulos de classe e os ramos representam conjunções de recursos que levam a esses rótulos de classe. As árvores de decisão nas quais a variável de destino pode assumir valores contínuos (geralmente números reais) são chamadas de árvores de regressão.

In [ ]:
# Decision Tree
 
decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, Y_train)
Y_pred = decision_tree.predict(X_test)
acc_decision_tree = round(decision_tree.score(X_train, Y_train) * 100, 2)
acc_decision_tree

 O próximo modelo Random Forests é um dos mais populares. Florestas aleatórias ou florestas de decisão aleatória são um método de aprendizado conjunto para classificação, regressão e outras tarefas, que operam construindo uma infinidade de árvores de decisão (n_estimators = 100) no momento do treinamento e gerando a classe que é o modo das classes (classificação) ou previsão média (regressão) das árvores individuais.

In [ ]:
# Random Forest
 
random_forest = RandomForestClassifier(n_estimators=100)
random_forest.fit(X_train, Y_train)
Y_pred = random_forest.predict(X_test)
random_forest.score(X_train, Y_train)
acc_random_forest = round(random_forest.score(X_train, Y_train) * 100, 2)
acc_random_forest

 ### Avaliação do modelo
Agora podemos classificar nossa avaliação de todos os modelos para escolher o melhor para o nosso problema. Embora a Árvore de Decisão e a Floresta Aleatória tenham a mesma pontuação, optamos por usar a Floresta Aleatória, pois elas corrigem o hábito de adaptação das árvores de decisão ao conjunto de treinamento.

In [ ]:
models = pd.DataFrame({
    'Model': ['Support Vector Machines', 'KNN', 'Logistic Regression', 
              'Random Forest', 'Naive Bayes', 'Perceptron', 
              'Stochastic Gradient Decent', 'Linear SVC', 
              'Decision Tree'],
    'Score': [acc_svc, acc_knn, acc_log, 
              acc_random_forest, acc_gaussian, acc_perceptron, 
              acc_sgd, acc_linear_svc, acc_decision_tree]})
models.sort_values(by='Score', ascending=False)

In [ ]:
submission = pd.DataFrame({
        "PassengerId": test_df["PassengerId"],
        "Survived": Y_pred
    })
# submission.to_csv('../output/submission.csv', index=False)